File: app/src/main/java/quangnam/com/sample/base/gson/ResponseAdapterFactory.java
package quangnam.com.sample.base.gson;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.TypeAdapter;
import com.google.gson.TypeAdapterFactory;
import com.google.gson.reflect.TypeToken;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonWriter;

import java.io.IOException;

import quangnam.com.sample.data.exception.ApiException;
import quangnam.com.sample.data.network.response.BaseErrorResponse;
import quangnam.com.sample.data.network.response.ResponseWrapper;
import quangnam.com.sample.util.Constant;

/**
 * Created by quangnam on 2/26/18.
 * Project Sample
 */

public class ResponseAdapterFactory implements TypeAdapterFactory {


    @Override
    public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> type) {
        final TypeAdapter<T> delegate = gson.getDelegateAdapter(this, type);
        final TypeAdapter<JsonElement> elementTypeAdapter = gson.getAdapter(JsonElement.class);
        TypeToken<ResponseWrapper<BaseErrorResponse>> token = new TypeToken<ResponseWrapper<BaseErrorResponse>>(){};
        final TypeAdapter<ResponseWrapper<BaseErrorResponse>> errorAdapter = gson.getDelegateAdapter(this, token);

        if (!type.getRawType().isAssignableFrom(ResponseWrapper.class)) {
            return delegate;
        }

        return new TypeAdapter<T>() {
            @Override
            public void write(JsonWriter out, T value) throws IOException {
                delegate.write(out, value);
            }

            @Override
            public T read(JsonReader in) throws IOException {
                JsonElement element = elementTypeAdapter.read(in);

                if (element.isJsonObject()) {
                    JsonObject object = element.getAsJsonObject();
                    if (object.has(ResponseWrapper.STATUS)) {
                        String status = object.get(ResponseWrapper.STATUS).getAsString();
                        if (!status.equalsIgnoreCase(Constant.ResponseStatus.success)) {
                            ResponseWrapper<BaseErrorResponse> response = errorAdapter.fromJsonTree(element);
                            throw new ApiException(Integer.parseInt(response.getCode()),
                                    response.getObject().getValue());
                        }
                    }
                }

                return delegate.fromJsonTree(element);
            }
        };
    }
}

File: app/src/main/java/quangnam/com/sample/base/gson/ResponseAdapterFactory.kt
package quangnam.com.sample.base.gson

import com.google.gson.Gson
import com.google.gson.JsonElement
import com.google.gson.TypeAdapter
import com.google.gson.TypeAdapterFactory
import com.google.gson.reflect.TypeToken
import com.google.gson.stream.JsonReader
import com.google.gson.stream.JsonWriter
import quangnam.com.sample.data.exception.ApiException
import quangnam.com.sample.data.network.response.BaseErrorResponse
import quangnam.com.sample.data.network.response.ResponseWrapper
import quangnam.com.sample.util.AppUtils
import quangnam.com.sample.util.Constant
import java.io.IOException
import javax.inject.Inject

/**
 * Created by quangnam on 2/26/18.
 * Project Sample
 */

class ResponseAdapterFactory @Inject
internal constructor(private val mAppUtil: AppUtils) : TypeAdapterFactory {

    override fun <T> create(gson: Gson, type: TypeToken<T>): TypeAdapter<T> {
        val delegate = gson.getDelegateAdapter(this, type)
        val elementTypeAdapter = gson.getAdapter(JsonElement::class.java)
        val token = object : TypeToken<ResponseWrapper<BaseErrorResponse>>() {

        }
        val errorAdapter = gson.getDelegateAdapter(this, token)

        return if (!type.rawType.isAssignableFrom(ResponseWrapper::class.java)) {
            delegate
        } else object : TypeAdapter<T>() {
            @Throws(IOException::class)
            override fun write(out: JsonWriter, value: T) {
                delegate.write(out, value)
            }

            @Throws(IOException::class)
            override fun read(`in`: JsonReader): T {
                val element = elementTypeAdapter.read(`in`)

                if (element.isJsonObject) {
                    val `object` = element.asJsonObject
                    if (`object`.has(ResponseWrapper.STATUS)) {
                        val status = `object`.get(ResponseWrapper.STATUS).asString
                        if (!status.equals(Constant.ResponseStatus.success, ignoreCase = true)) {
                            val response = errorAdapter.fromJsonTree(element)
                            throw mAppUtil.addLocalizedException(ApiException(Integer.parseInt(response.code), response.`object`!!.value ?: ""))
                        }
                    }
                }

                return delegate.fromJsonTree(element)
            }
        }

    }
}
